docs/en/week13/13-3.md

```diff
@@ -0,0 +1,518 @@
+        batch_x = batch_graphs.ndata['feat']
+        batch_e = batch_graphs.edata['feat']
+        
+        batch_scores = model(batch_graphs, batch_x, batch_e, batch_snorm_n, batch_snorm_e)
+        J = loss(batch_scores, batch_labels)
+        optimizer.zero_grad()
+        J.backward()
+        optimizer.step()
+        
+        epoch_loss += J.detach().item()
+        epoch_train_acc += accuracy(batch_scores, batch_labels)
+        nb_data += batch_labels.size(0)
+
+    epoch_loss /= (iter + 1)
+    epoch_train_acc /= nb_data
+
+    return epoch_loss, epoch_train_acc
+```
+In our `train` function, we have our generic code which takes samples from `dataloader`.  Next, `batch_graphs`, `batch_x`, `batch_e`, `batch_snorm_n` and `batch_snorm_e` are fed into our model which returns `batch_scores` (of size 8). The predicted scores are compared with the ground truth in our loss function: `loss(batch_scores, batch_labels)`. Then, we zero out gradients (`optimizer.zero_grad()`), perform backprop (`J.backward()`) and updates our weights (`optimizer.step()`). Finally, loss for epoch and training accuracy is calculated. 
+
+We use a similar code for our `evaluate` function as shown below: 
+```python
+def evaluate(model, data_loader, loss):
+
+    model.eval()
+    epoch_test_loss = 0
+    epoch_test_acc = 0
+    nb_data = 0
+    
+    with torch.no_grad():
+        for iter, (batch_graphs, batch_labels, batch_snorm_n, batch_snorm_e) in enumerate(data_loader):
+            batch_x = batch_graphs.ndata['feat']
+            batch_e = batch_graphs.edata['feat']
+
+            batch_scores = model(batch_graphs, batch_x, batch_e, batch_snorm_n, batch_snorm_e)
+            J = loss(batch_scores, batch_labels)
+
+            epoch_test_loss += J.detach().item()
+            epoch_test_acc += accuracy(batch_scores, batch_labels)
+            nb_data += batch_labels.size(0)
+
+        epoch_test_loss /= (iter + 1)
+        epoch_test_acc /= nb_data
+    
+    return epoch_test_loss, epoch_test_acc
+```
+
+Finally, we are ready to train! We load train/test data, instantiate our model, define our loss and optimizer, and start training as shown below:
+```python
+
+# datasets
+train_loader = DataLoader(trainset, batch_size=50, shuffle=True, collate_fn=collate)
+test_loader = DataLoader(testset, batch_size=50, shuffle=False, collate_fn=collate)
+
+# Create model
+model = GatedGCN(input_dim=1, hidden_dim=100, output_dim=8, L=4)
+loss = nn.CrossEntropyLoss()
+optimizer = torch.optim.Adam(model.parameters(), lr=0.0001)
+
+epoch_train_losses = []
+epoch_test_losses = []
+epoch_train_accs = []
+epoch_test_accs = []
+
+for epoch in range(40):
+    
+    start = time.time()
+    train_loss, train_acc = train(model, train_loader, loss)
+    test_loss, test_acc = evaluate(model, test_loader, loss)
+        
+    print(f'Epoch {epoch}, train_loss: {train_loss:.4f}, test_loss: {test_loss:.4f}')
+    print(f'train_acc: {train_acc:.4f}, test_acc: {test_acc:.4f}')
+
```

Comment:
Nothing special here. People **do** have access to the notebook.

Reply:
removed